### ISSUE TITLE
Allow disabling MFA with a yaml setting

### ISSUE THREAD

**Issue by ScopeyNZ** [silverstripe/silverstripe-mfa#145]
### Description

It's not inconceivable that in dev and test environments you might wish to disable MFA so that your users aren't burdened. Additionally, with methods like WebAuthn and any that use the encryption service, these methods won't transfer well from database snapshots as they rely on environment specific variables.

### A/Cs

- I can set an environment setting via `yaml` to disable MFA in specific environments
- MFA is not prompted for verification or registration - but not removing the functionality entirely.

### Notes

It might make sense to still allow setting up MFA - although it won't be required for any verification.

**Comment by chillu** [comment]
As an alternative approach, we could turn it off via module logic - untick and disable the "require MFA" checkbox in the CMS. Add a warning with something like "to enable MFA on non-production environments, get your dev to set this flag in code/env. Warning: This will break existing logins imported from other environments. You can always use the default admin to login to the CMS and reset logins".

**Comment by ScopeyNZ** [comment]
cc @silverstripeux

**Comment by brynwhyman** [comment]
> untick and disable the "require MFA" checkbox in the CMS on non-prod envs

Just pointing out that this conflicts with our recent discussion about removing the ability to disable MFA from the CMS. See #144

**Comment by brynwhyman** [comment]
This sounds like a solution to https://github.com/silverstripeltd/cc-issues/issues/315

Be careful not to split these up and loose sight of the ACs

**Comment by chillu** [comment]
> Just pointing out that this conflicts with our recent discussion about removing the ability to disable MFA from the CMS

I think that's a good default for prod envs, but for non-prod envs the reality is that we can't have it all. Either you can enforce MFA, and have all logins broken by default when copying a prod snapshot until an admin fixes them user-by-user. Or you can disable MFA on non-prod, have logins work with just passwords as-is (even when those passwords are copied from a prod snapshot).

**Comment by ScopeyNZ** [comment]
We'll probably have to add a task for a developer to run that can clear out existing MFA configuration on Members? Say they wanted to enable MFA on UAT after restoring a prod db snapshot - they'll have to clear out the `MFARegisteredMethod` table too.

**Comment by newleeland** [comment]
> As an alternative approach, we could turn it off via module logic - untick and disable the "require MFA" checkbox in the CMS on non-prod envs. Add a warning with something like "to enable MFA on non-production environments, get your dev to set this flag in code/env. Warning: This will break existing logins imported from other environments. You can always use the default admin to login to the CMS and reset logins".

A warning could work on non-prod environments (under security in CMS), if we can't solve the issue of copying a prod snapshot to non-prod without breaking logins. So if MFA has been enabled on UAT by the developer, would it mean the UAT logins would break everytime a new snapshot from prod is copied over to uat?

We would also need to consider dev docs work with this.

**Comment by brynwhyman** [comment]
During our grooming session we discussed that this would make more sense at a yaml config level so that it could be more prescriptive, rather than just dev or not-dev.

It also means that this could be implemented irrespective of https://github.com/silverstripeltd/cc-issues/issues/315

**Comment by robbieaverill** [comment]
Can we clarify whether this means that MFA is disabled entirely, as it would be had you not installed the module, or can you still access the MFA endpoints and see the CMS fields on your profile, but the redirect to registration or verification in user flows simply won't kick in?

**Comment by robbieaverill** [comment]
@newleeland @brynwhyman we still need to clarify the expected behaviour of this before we can implement it

**Comment by ScopeyNZ** [comment]
Just to clarify.

We make yaml config to disable MFA. I assume this means there's no indication of MFA at all.

Do we still want to provide some UI indication that MFA is disabled for instance where it's enabled on production but not on UAT (eg. https://projects.invisionapp.com/share/3PNSKZQYBJZ#/screens/366124954). Perhaps this can be a separate yaml configuration?

If the default is "off" on non-prod environments - I think there's real value in having a message for developers and users who might be wondering where their MFA is gone.

Additionally, do we have any task to capture resetting/clearing MFA settings when a database snapshot is deployed? We'll need this if a developer needs to turn on MFA for a non-production environment after restoring a database snapshot because users will be asked for credentials but none of them will work.

**Comment by brynwhyman** [comment]
> If the default is "off" on non-prod environments

I'm not aware on implementing "off by default" for non-prod environments, nor does that seem like a good idea? This issue would mostly be to support using local environments that don't hold production data.

The AC states the expected behaviour: "MFA is not prompted for verification or registration - but not removing the functionality entirely."

> Additionally, do we have any task to capture resetting/clearing MFA settings when a database snapshot is deployed? We'll need this if a developer needs to turn on MFA for a non-production environment after restoring a database snapshot because users will be asked for credentials but none of them will work.

@ScopeyNZ, see https://github.com/silverstripeltd/cc-issues/issues/315#issuecomment-501056959. We don't have a issue capturing that yet, although I'd appreciate it if you raised one! I'd expect this is mostly mitigated with this https://github.com/silverstripeltd/cc-issues/issues/315#issuecomment-501056959. Is that correct?

**Comment by robbieaverill** [comment]
Thanks, in that case presumably you can still access MFA information if you want, but the redirects won't happen on login

**Comment by robbieaverill** [comment]
PR at https://github.com/silverstripe/silverstripe-mfa/pull/193

**Comment by ScopeyNZ** [comment]
> I'm not aware on implementing "off by default" for non-prod environments

Sorry I was under the assumption that the common use case was _not_ sharing the same key on production/UAT envs. I was out of the loop on recent discussions I didn't realise that changed. You can ignore me.
